src/inmanta/server/services/environmentservice.py

```diff
@@ -194,6 +197,18 @@ async def delete_environment(self, environment_id: uuid.UUID) -> Apireturn:
         await self.environment_delete(environment_id)
         return 200
 
+    @protocol.handle(methods.halt_environment, env="tid")
+    async def halt(self, env: data.Environment) -> None:
+        await env.update_fields(halted=True)
+        await self.agent_manager.halt_agents(env)
+        await self.autostarted_agent_manager.stop_agents(env)
```

Comment:
Ensure this method cannot be called when the environment is already halted. This may corrupt the values in the `unpause_on_resume` column.